kotlinx-coroutines-core/wasmWasi/src/internal/EventLoop.kt

```diff
@@ -0,0 +1,206 @@
+@file:OptIn(UnsafeWasmMemoryApi::class)
+
+package kotlinx.coroutines.internal
+
+import kotlinx.coroutines.*
+import kotlin.math.min
+import kotlin.wasm.WasmImport
+import kotlin.wasm.unsafe.MemoryAllocator
+import kotlin.wasm.unsafe.Pointer
+import kotlin.wasm.unsafe.UnsafeWasmMemoryApi
+import kotlin.wasm.unsafe.withScopedMemoryAllocator
+
+@WasmImport("wasi_snapshot_preview1", "poll_oneoff")
+private external fun wasiPollOneOff(ptrToSubscription: Int, eventPtr: Int, nsubscriptions: Int, resultPtr: Int): Int
+
+@WasmImport("wasi_snapshot_preview1", "clock_time_get")
+private external fun wasiRawClockTimeGet(clockId: Int, precision: Long, resultPtr: Int): Int
+
+private const val CLOCKID_MONOTONIC = 1
+
+private class Event(var callback: (() -> Unit)?, val absoluteTimeout: Long): DisposableHandle {
+    companion object {
+        fun createCancelled(): Event = Event(null, 0)
+    }
+
+    override fun dispose() {
+        if (callback == null) return
+        callback = null
+        nextCycleNearestEventAbsoluteTime = 0
+        nextCycleContainTimedEvent = true
+    }
+}
+
+private var currentCycleEvents = mutableListOf<Event>()
+private var nextCycleEvents = mutableListOf<Event>()
+private var thrownException: Throwable? = null
+private var nextCycleNearestEventAbsoluteTime: Long = Long.MAX_VALUE
+private var nextCycleContainTimedEvent = false
+
+private fun initializeSubscriptionPtr(allocator: MemoryAllocator): Pointer {
+    val ptrToSubscription = allocator.allocate(48)
+    //userdata
+    ptrToSubscription.storeLong(0)
+    //uint8_t tag;
+    (ptrToSubscription + 8).storeByte(0) //EVENTTYPE_CLOCK
+    //__wasi_clockid_t id;
+    (ptrToSubscription + 16).storeInt(CLOCKID_MONOTONIC) //CLOCKID_MONOTONIC
+    //__wasi_timestamp_t timeout;
+    //(ptrToSubscription + 24).storeLong(timeout)
```

Comment:
Why is this commented out? The documentation of `allocate` states that it returns "uninitialized" memory; does it mean that the memory may contain garbage?
* If yes, we need to store the timeout;
* If not, what is the reason to explicitly write `0` in the other places in the same function? It looks like we have exactly one non-zero bit, the clock id.

Reply:
It returns `uninitialized` memory and but waiting time loaded in `sleep` function.
All fields in this structure are the same between `sleep` calls but this `timeout` field. We load them once here and load the `timeout` in `sleep` function calls later.